### PR TITLE
[Driver] Report invalid target triple versions for all environment types.

### PR DESCRIPTION
**Pull request by ZijunZhaoCCK**
Followup for https://github.com/llvm/llvm-project/pull/75373

1. Make this feature not just available for android, but everyone.
2. Correct some target triples.
3. Add opencl to the environment type list.

### CODE REVIEW

File: clang/test/Frontend/fixed_point_bit_widths.c [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // RUN: %clang -x c -ffixed-point -S -emit-llvm -o - %s | FileCheck %s
-// RUN: %clang -x c -ffixed-point -S -emit-llvm -o - --target=x86_64-scei-ps4-ubuntu-fast %s | FileCheck %s
```
Comment (MaskRay) [inline clang/test/Frontend/fixed_point_bit_widths.c:2]: @pogo59

Comment (pogo59) [reply clang/test/Frontend/fixed_point_bit_widths.c:2]: Thanks for tagging me. These are definitely invalid triples. It appears they were added in f921d854, which was a recommit of db55d833, where the test was introduced. I don't see that they are really adding anything to the coverage, so I think it's fine to remove them.

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -1443,15 +1443,17 @@ Compilation *Driver::BuildCompilation(ArrayRef<const char *> ArgList) {
...
-      Diags.Report(diag::err_drv_triple_version_invalid)
-          << TripleVersionName << TC.getTripleString();
-      ContainsError = true;
-    }
+  // Check if the environment version is valid.
+  llvm::Triple Triple = TC.getTriple();
+  StringRef TripleVersionName = Triple.getEnvironmentVersionString();
+  StringRef TripleObjectFormat =
+      Triple.getObjectFormatTypeName(Triple.getObjectFormat());
+
```
Comment (MaskRay) [inline clang/lib/Driver/Driver.cpp:1451]: (the prevailing code style does not insert a blank line in this case. )

File: llvm/include/llvm/TargetParser/Triple.h [hunk 3]
```diff
@@ -255,7 +255,7 @@ class Triple {
     Cygnus,
     CoreCLR,
     Simulator, // Simulator variants of other systems, e.g., Apple's iOS
-    MacABI, // Mac Catalyst variant of Apple's iOS deployment target.
+    MacABI,    // Mac Catalyst variant of Apple's iOS deployment target.
```
Comment (MaskRay) [inline llvm/include/llvm/TargetParser/Triple.h:258]: Revert this difference? You can ignore clang-format reports.

File: llvm/include/llvm/TargetParser/Triple.h [hunk 4]
```diff
@@ -276,7 +276,7 @@ class Triple {
     Callable,
     Mesh,
     Amplification,
-
+    OpenCL,
```
Comment (MaskRay) [inline llvm/include/llvm/TargetParser/Triple.h:279]: I wonder why we need this addition. This is not mentioned in the description.

Comment (ZijunZhaoCCK) [reply llvm/include/llvm/TargetParser/Triple.h:279]: Some cases like https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenOpenCL/amdgpu-alignment.cl#L3

File: llvm/lib/TargetParser/Triple.cpp [hunk 5]
```diff
@@ -1219,8 +1222,24 @@ VersionTuple Triple::getEnvironmentVersion() const {
...
+
   StringRef EnvironmentTypeName = getEnvironmentTypeName(getEnvironment());
   EnvironmentName.consume_front(EnvironmentTypeName);
+
+  if (EnvironmentName.contains("-")) {
+    // -obj is the suffix
+    if (getObjectFormat() != Triple::UnknownObjectFormat) {
+      StringRef ObjectFormatTypeName =
+          getObjectFormatTypeName(getObjectFormat());
+      StringRef Suffix = (Twine("-") + ObjectFormatTypeName).str();
```
Comment (MaskRay) [inline llvm/lib/TargetParser/Triple.cpp:1239]: This creates a temporary `std::string` and binds it on a StringRef. The temporary `std::string` is then destroyed, leading to a dangling reference, which Clang helpfully warns -Wdangling-gsl

File: clang/test/Driver/mips-features.c [hunk 6]
```diff
@@ -400,12 +400,12 @@
 // LONG-CALLS-DEF-NOT: "long-calls"
 //
 // -mbranch-likely
-// RUN: %clang -target -mips-mti-linux-gnu -### -c %s -mbranch-likely 2>&1 \
+// RUN: %clang -target mips-mti-linux-gnu -### -c %s -mbranch-likely 2>&1 \
```
Comment (MaskRay) [inline clang/test/Driver/mips-features.c:403]: while updating, switching to `--target=`. `-target ` has been deprecated since Clang 3.4

**Comment by MaskRay** [comment]
If I remove `OpenCL`, `clang/test/CodeGenOpenCL/amdgpu-alignment.cl` won't fail. It seems that OpenCL is a not a necessary change in this PR.
```
--- i/llvm/lib/TargetParser/Triple.cpp
+++ w/llvm/lib/TargetParser/Triple.cpp
@@ -325,4 +325,2 @@ StringRef Triple::getEnvironmentTypeName(EnvironmentType Kind) {
   case Amplification: return "amplification";
-  case OpenCL:
-    return "opencl";
   case OpenHOS: return "ohos";
@@ -696,3 +694,2 @@ static Triple::EnvironmentType parseEnvironment(StringRef EnvironmentName) {
       .StartsWith("amplification", Triple::Amplification)
-      .StartsWith("opencl", Triple::OpenCL)
       .StartsWith("ohos", Triple::OpenHOS)
```

**Comment by ZijunZhaoCCK** [comment]
> If I remove `OpenCL`, `clang/test/CodeGenOpenCL/amdgpu-alignment.cl` won't fail. It seems that OpenCL is a not a necessary change in this PR.
> 
> ```
> --- i/llvm/lib/TargetParser/Triple.cpp
> +++ w/llvm/lib/TargetParser/Triple.cpp
> @@ -325,4 +325,2 @@ StringRef Triple::getEnvironmentTypeName(EnvironmentType Kind) {
>    case Amplification: return "amplification";
> -  case OpenCL:
> -    return "opencl";
>    case OpenHOS: return "ohos";
> @@ -696,3 +694,2 @@ static Triple::EnvironmentType parseEnvironment(StringRef EnvironmentName) {
>        .StartsWith("amplification", Triple::Amplification)
> -      .StartsWith("opencl", Triple::OpenCL)
>        .StartsWith("ohos", Triple::OpenHOS)
> ```

Oh , my bad. Failed tests are ```Failed Tests (4):
  Clang :: CodeGenOpenCL/amdgpu-debug-info-pointer-address-space.cl
  Clang :: CodeGenOpenCL/amdgpu-debug-info-variable-expression.cl
  Clang :: Driver/amdgpu-toolchain-opencl.cl
  Clang :: Driver/opencl.cl```

File: llvm/lib/TargetParser/Triple.cpp [hunk 7]
```diff
@@ -1219,8 +1222,25 @@ VersionTuple Triple::getEnvironmentVersion() const {
...
+
   StringRef EnvironmentTypeName = getEnvironmentTypeName(getEnvironment());
   EnvironmentName.consume_front(EnvironmentTypeName);
+
+  if (EnvironmentName.contains("-")) {
+    // -obj is the suffix
+    if (getObjectFormat() != Triple::UnknownObjectFormat) {
+      StringRef ObjectFormatTypeName =
+          getObjectFormatTypeName(getObjectFormat());
+      const std::string &tmp = (Twine("-") + ObjectFormatTypeName).str();
```
Comment (MaskRay) [inline llvm/lib/TargetParser/Triple.cpp:1239]: Just use `const std::string`. The lifetime extension due to `const &` is subtle and fragile, and does not give any benefit here.

https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary
https://abseil.io/tips/107

Then `EnvironmentName.consume_back(tmp);`

Avoid the used-once `StringRef Suffix = tmp;`

**Review by MaskRay — APPROVED** [review]
A better title may be:

[Driver] Report invalid target triple versions for all environment types.

"Driver" is better than "clang" as it is specific about where the error arises.

**Review by pirama-arumuga-nainar — APPROVED** [review]

**Comment by glandium** [comment]
This broke the wasi-threads target:
`clang: error: version 'threads' in target triple 'wasm32-unknown-wasi-threads' is invalid`

**Comment by ZijunZhaoCCK** [comment]
> This broke the wasi-threads target: `clang: error: version 'threads' in target triple 'wasm32-unknown-wasi-threads' is invalid`

Because `threads` is not in EnvironmentType list: https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/TargetParser/Triple.h#L231 or ObjectType list: https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/TargetParser/Triple.h#L284 . 

The format should be `arch-vendor-os-env`. If `threads` is a new environment type, please add it in the list. If `wasi-threads` is a special case, please let me know!

And one more question, would you mind pointing me out the test case of `wasm32-unknown-wasi-threads` ? I don't see it. Thank you!

**Comment by glandium** [comment]
We stumbled upon this downstream because we have jobs building wasi-sdk with clang-trunk, and wasi-sdk builds some things with that target. It apparently comes from https://github.com/WebAssembly/wasi-libc/pull/381

**Comment by glandium** [comment]
There's apparently also wasm32-wasi-preview2 and wasm32-wasi-pthread, which I suppose are equally broken by this change.

**Comment by ZijunZhaoCCK** [comment]
> There's apparently also wasm32-wasi-preview2 and wasm32-wasi-pthread, which I suppose are equally broken by this change.

Yes, I think so. I think adding these environment types in wasi-libc repo could help fix those errors.

**Comment by glandium** [comment]
> Yes, I think so. I think adding these environment types in wasi-libc repo could help fix those errors.

WDYM?

**Comment by MaskRay** [comment]
> > There's apparently also wasm32-wasi-preview2 and wasm32-wasi-pthread, which I suppose are equally broken by this change.
> 
> Yes, I think so. I think adding these environment types in wasi-libc repo could help fix those errors.

If wasm can arbitrary environment types, it seems that we can opt out the check for `isWasm()`.
We define new environment types when they affect the compiler behavior. If wasm arbitrary sets the environment, we should not define a type for each one.

**Comment by ZijunZhaoCCK** [comment]
> > > There's apparently also wasm32-wasi-preview2 and wasm32-wasi-pthread, which I suppose are equally broken by this change.
> > 
> > 
> > Yes, I think so. I think adding these environment types in wasi-libc repo could help fix those errors.
> 
> If wasm can arbitrary environment types, it seems that we can opt out the check for `isWasm()`. We define new environment types when they affect the compiler behavior. If wasm arbitrary sets the environment, we should not define a type for each one.

Okay, I will add this check.

**Comment by MrSidims** [comment]
Hi @ZijunZhaoCCK , sorry, I'm quite late for the party, but may I ask you some questions?

We want to be able to distinguish between Vulkan and OpenCL environments in [SPIR-V backend](https://github.com/intel/llvm/tree/sycl/llvm/lib/Target/SPIRV), so we could be able to pick between Shader and Kernel capabilities depending on the environment. And OpenCL environment added by this patch seems like the perfect match for us.

We would also need to change OpenCL CodeGen in clang to start emitting this part of the triple, but before doing that I'd like to check, what was the intention under adding OpenCL environment in this PR (as currently it's not exposed in LLVM anywhere, so I could take a look)? Particularly I'd like to know, whether your intentions were to add OpenCL as a language to the triple or OpenCL as an execution environment?

**Pull request merged**